### ISSUE TITLE
conserve X-Poedit-* headers with --merge

### ISSUE THREAD

**Issue by drzraf** [wp-cli/i18n-command#50]
When using: `wp i18n make-pot . languages/theme.pot --merge` on a pot file already created with POedit, I'd like to keep custom headers since they contain many useful hints for contributors (possibly using poedit).
Eg:
```
"X-Poedit-Basepath: ..\n"
"X-Poedit-KeywordsList: __;_e;_n;_x;_ex;_nx;esc_attr__;esc_attr_e;esc_attr_x;esc_html__;esc_html_e;esc_html_x;_n_noop;_nx_noop;translate_nooped_plural\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"
"X-Poedit-SearchPath-0: .\n"
"X-Poedit-SearchPathExcluded-0: foo\n"
"X-Poedit-SearchPathExcluded-1: bar.php\n"
```

**Comment by schlessera** [comment]
Yes, this makes sense to me.

I'll defer the review of the PR to @swissspidy, as he's the expert in that domain.

**Comment by swissspidy** [comment]
Related Twitter thread: https://twitter.com/Mte90Net/status/1014512091036897281

I'm curious to know what exactly Poedit does with these headers. Are they really useful if you don't need to generate a POT file using Poedit, but only need to translate?

If so, we should correctly add these headers on our own, instead of just copying the ones from an existing POT file.

In your scenario, the POT file created by WP-CLI is the single source of truth, as it contains all translatable strings. If anything, WP-CLI should declare which paths were excluded (or perhaps combine that header), what the base path is, and which functions are scanned.

Also, the "Plural-Forms" header doesn't make for a POT file as it contains no language information and thus no plural form information.

**Comment by drzraf** [comment]
There are useful since they preload the configuration used by Poedit when it scans project's directories.
Avoiding repetitive, error-prone, **re**configuration each time regenerated po files are loaded by Poedit is useful.

**Comment by swissspidy** [comment]
Why do you need Poedit to scan the project directory when you already use WP-CLI to do that and create a POT file?

When you create a POT file using WP-CLI, all you need to do is create a new translation in Poedit based on that catalog. I don't see how Poedit needs these headers in that case.

**Comment by drzraf** [comment]
Because pot file "may" (will) be used by people relying on Poedit for
some tasks.
There are no uniform way to manage collaborative translationsand
hard to impose a workflow (especially for new project). It's preferible
to not lock any door (including poedit)...
(IHMO even if that means preserving a couple of non-standard headers
lines in the file)

**Comment by swissspidy** [comment]
I agree that we shouldn't lock any doors, but can you explain a bit on these tasks people (might) rely on Poedit for? Just to help me properly understand it :-)

When using this WP-CLI command, your use case should be:

1. Create POT file using `wp i18n make-pot` which scans your whole project for translatable strings.
2. Give this POT file to your translators.
3. Translator opens Poedit and creates a new translation from that POT file and saves it as de_DE.po or so.

`X-Poedit-Basepath`, `X-Poedit-KeywordsList`, `X-Poedit-SearchPath` and `X-Poedit-SearchPathExcluded` are only needed if you use Poedit to scan your project for translatable strings (step 1 above). Source: https://github.com/vslavik/poedit/wiki/PO-Extensions.

However, when you already use WP-CLI for that, these are irrelevant. Why would you as a developer want to re-run string extraction in Poedit when WP-CLI already does that for you? And as a translator you don't want to do this anyway.

In other words:

In theory, we can just add `X-Poedit-Basepath`, `X-Poedit-KeywordsList`, `X-Poedit-SearchPath` and `X-Poedit-SearchPathExcluded` headers in this command, but a) I see no real value in it and b) I wouldn't copy these from the existing POT file that is being merged using `--merge`.

Interestingly, @mte90 built https://github.com/CodeAtCode/PoEditZ, a script which uses the `X-Poedit-SearchPathExcluded` header from an existing POT file and feeds this information to `--exclude`. However, I don't think this is wanted in all cases.

**Comment by swissspidy** [comment]
Looking at https://github.com/vslavik/poedit/wiki/PO-Extensions again, the only headers I find worth adding are `X-Generator` and `X-Source-Language`. These aren't used for string extraction by Poedit, but could help developers and translators, respectively.

I'll open new issues for these.

**Comment by drzraf** [comment]
But why keeping Poedit from (fully) working (including new extractions, ie `KeywordsList`, ...) when we could continue fully supporting it in a compatible fashion (just conserving existing header)?
Should `make-pot` really act as be a po validator?
With `--merge` shouldn't it rather try to respect existing files as much as possible?

**Comment by drzraf** [comment]
On the theoretical side: Why keeping Poedit from (fully) working (including new extractions, ie `KeywordsList`, ...) when we could continue fully supporting it in a compatible fashion (just conserving existing header)? Should `make-pot` really act as be a po validator? With `--merge` shouldn't it rather try to respect existing files as much as possible? Preexisting `X-` headers, although note officially part of the format, are legitimate.

On the practical side: I want `make-pot` for use with automation, `zanata push/pull`...
But the same way not all translators may tool X (be it Zanata or any other), not all developers are fine with running `wp-cli make-pot`. If one developer prefer to refresh strings using poedit I'd like to "tolerate". I don't want to drop its local configuration when I refresh and commit using `make-pot` myself.

**Comment by Mte90** [comment]
As explained by Pascal my idea was to keep the compatibility with PoEdit and use their headers with this tool.
I removed the support with the other parameters but my script keep the other PoEdit headers in the new po file but they are not used from wp-cli because they are needed.

**Comment by swissspidy** [comment]
I'd much rather like to describe this as "Add Poedit compatible headers" instead of "Conser ve Poedit headers from existing file".

This way one _**developer**_ in your team can still use Poedit, and and other can use WP-CLI.

(I'd recommend against doing that as they might not lead to the same result).

As for _**translators**_, the Poedit headers aren't relevant.

> With `--merge` shouldn't it rather try to respect existing files as much as possible?

Let's say you have an existing POT file called `foo.pot` with the following headers:

```po
"X-Poedit-SearchPathExcluded-0: foo\n"
"X-Poedit-SearchPathExcluded-1: bar.php\n"
```

When I run `wp i18n make-pot --merge=foo.pot --exclude=myfolder`, I'd expect either

```po
"X-Poedit-SearchPathExcluded-0: myfolder\n"
```

(plus anything from https://github.com/wp-cli/i18n-command/blob/bd16b5fd631980dea4684e9b290b3e80674b4401/src/MakePotCommand.php#L39)

or perhaps

```po
"X-Poedit-SearchPathExcluded-0: foo\n"
"X-Poedit-SearchPathExcluded-1: bar.php\n"
"X-Poedit-SearchPathExcluded-2: myfolder\n"
```

(plus anything from https://github.com/wp-cli/i18n-command/blob/bd16b5fd631980dea4684e9b290b3e80674b4401/src/MakePotCommand.php#L39)

as a result.

WP-CLI creates a completely new POT file. It shouldn't just copy Poedit headers from an existing POT file 1:1. Instead, I'd much rather add `X-Poedit-Basepath`, `X-Poedit-KeywordsList`, `X-Poedit-SearchPath` on our own.

Let's say you have an existing POT file called `foo.pot` with the following headers:

```po
"X-Poedit-Basepath: ..\n"
```

When I run `wp i18n make-pot --merge=foo.pot mypluginfolder/ mypluginfolder/assets/languages/result.pot`

I expect `X-Poedit-Basepath` to be `../../`, not `..`.

Again, that's why I would not just copy Poedit headers but create them on our own.

**Comment by vslavik** [comment]
Poedit developer here. @swissspidy asked me to chip in on this issue, but there's little need: he's right about everything. These headers are for extraction only and I'd say it's actually contra-productive to preserve (or add) them if a different tool should be used.

`X-Source-Language` is only useful if source code's text isn't in English, I don't know how common that is (I typically see such things with non-WP files). And even then it's only useful if there's a confusing mix with English, otherwise Poedit detects the source language automatically.

`X-Generator` is useful and is a convention used by others, not only Poedit.

**Comment by swissspidy** [comment]
Thanks a lot for chiming in!

This confirms my standpoint that it doesn't make sense to add this to WP-CLI.

`X-Source-Language` can be handled by #47 and `X-Generator` by #53.

I am closing this for now.
